### ISSUE TITLE
A potential bug in reward-model learning of summarize_rlhf

### ISSUE THREAD

**Issue by Shentao-YANG** [CarperAI/trlx#308]
### 🐛 Describe the bug

It seems to me that the reward-learning loss at `examples/summarize_rlhf/reward_model/reward_model.py` is wrong.

The current implementation is https://github.com/CarperAI/trlx/blob/724b6183202aa013a861efe723ba9f3dfa6664bc/examples/summarize_rlhf/reward_model/reward_model.py#L90
AFAIK, both `c_truncated_reward` and `r_truncated_reward` are tensors containing multiple "reward" values.

***By contrast***, in the original OpenAI paper, the loss is $loss(r_\theta) = -E_{(x, y_0, y_1, i) \sim D}[\log(\sigma(r_\theta(x, y_i) - r_\theta(x, y_{1-i})))]$, where there is only one reward for each candidate summary, rather than a tensor of rewards for partial summary. This may be implemented as 
```
loss += -torch.log(torch.sigmoid(chosen_rewards[i][c_ind] - rejected_rewards[i][r_ind])).mean()
```
My implementation also aligns with the comments in the same file 
https://github.com/CarperAI/trlx/blob/724b6183202aa013a861efe723ba9f3dfa6664bc/examples/summarize_rlhf/reward_model/reward_model.py#L59

Did I misunderstand something? If so, could you provide more explanation on the current reward-learning loss?

### Which trlX version are you using?

latest

**Comment by LouisCastricato** [comment]
@PhungVanDuy

**Comment by PhungVanDuy** [comment]
> ### 🐛 Describe the bug
> It seems to me that the reward-learning loss at `examples/summarize_rlhf/reward_model/reward_model.py` is wrong.
> 
> The current implementation is
> 
> https://github.com/CarperAI/trlx/blob/724b6183202aa013a861efe723ba9f3dfa6664bc/examples/summarize_rlhf/reward_model/reward_model.py#L90
> 
> 
> AFAIK, both `c_truncated_reward` and `r_truncated_reward` are tensors containing multiple "reward" values.
> _**By contrast**_, in the original OpenAI paper, the loss is loss(rθ)=−E(x,y0,y1,i)∼D[log⁡(σ(rθ(x,yi)−rθ(x,y1−i)))], where there is only one reward for each candidate summary, rather than a tensor of rewards for partial summary. This may be implemented as
> 
> ```
> loss += -torch.log(torch.sigmoid(chosen_rewards[i][c_ind] - rejected_rewards[i][r_ind])).mean()
> ```
> 
> My implementation also aligns with the comments in the same file
> 
> https://github.com/CarperAI/trlx/blob/724b6183202aa013a861efe723ba9f3dfa6664bc/examples/summarize_rlhf/reward_model/reward_model.py#L59
> 
> Did I misunderstand something? If so, could you provide more explanation on the current reward-learning loss?
> 
> ### Which trlX version are you using?
> latest
> 
> ### Additional system and package information
> _No response_

@Shentao-YANG  thank you for your concern. Actually, I already explained this before in this thread: https://github.com/CarperAI/trlx/issues/246#issuecomment-1413041378. The comment at line 59 will be removed, I just don't want to make a small PR for that, but sorry for this concern it will update in the PR this week. In terms of using pair in span calculation instead of the end position, I also explained here: https://github.com/CarperAI/trlx/issues/246#issuecomment-1414692881 (based on experiments result). 
If you are more interested in implementing a reward model I recommend you to check out [this](https://github.com/Dahoas/reward-modeling) repo.  There also provided options related to [pairwise](https://github.com/Dahoas/reward-modeling/blob/main/reward-modeling/finetune_rm.py#L40) as well as [dense](https://github.com/Dahoas/reward-modeling/blob/main/reward-modeling/finetune_rm.py#L28) for loss function.
